lib/tictactoe_ui.ex

```diff
@@ -1,88 +1,175 @@
-    get_tile_selection()
-    |> make_move(game_status)
-    |> decide_turn_type()
+  def get_player_move({_, board,_} = game_status) do
+    board_size = Board.size(board)
+    selection = get_tile_selection()
+    
+    if selection == :bad_input or selection > board_size do
+      do_turn_type(:bad_input, game_status)
+    else
+      make_move(selection, game_status)|> do_next_turn()
+    end
   end
 
-  def print_turn(:first_turn, game_status) do
+  def bad_input({player_symbol, _, _}= game_status) do
+    player = Markers.get_player_marker(player_symbol)
+    IO.puts @clear_screen
+    IO.puts "Bad input. Player #{player}, please re-enter selection."
+    print_board(game_status)
+    get_player_move(game_status)
+  end
+
+  def first_turn(game_status) do
     print_board(game_status)
     game_status
   end
 
-  def print_turn(:tile_already_selected, game_status) do
+  def tile_already_selected(game_status) do
     IO.puts "Tile already selected. Please select a different tile."
     print_board(game_status)
+    game_status
   end
 
-  def print_turn(:win, game_status) do
-    player_marker = get_marker_symbol(elem(game_status, 0))
+  def win(game_status) do
+    player_marker = get_marker_symbol(game_status)
     IO.puts "Player #{player_marker} has won the game."
     print_board(game_status)
     print_game_end_message
+    :game_over
   end
 
-  def print_turn(:tie, game_status) do
+  def tie(game_status) do
     IO.puts "This game is a tie."
     print_board(game_status)
     print_game_end_message
+    :game_over
   end
 
-  def print_game_end_message(), do: IO.puts "Goodbye."
+  defp print_game_end_message, do: IO.puts "Goodbye."
 
-  def print_board(game_status) do
-    elem(game_status, 1)
+  defp print_board({_, board, _}) do
+    board
     |> format_board()
     |> IO.write
-    game_status
+  end
+
+  defp get_tile_selection, do: IO.gets("Please enter a tile selection: ")|> check_tile_selection()
+
+  def check_tile_selection(selection) do
+    try do
+      String.trim(selection)
+      |> String.to_integer()
+      |> selection_is_negative()
+    rescue
+      ArgumentError -> :bad_input
+      FunctionClauseError -> :bad_input
+    end
+  end
+
+  defp selection_is_negative(selection) do
+    if selection <= 0 do
+      :bad_input
+    else
+      selection
+    end
   end
 
   def format_board(board) do
-    insert_new_lines_into_board([], board)
+    board
+    |> add_spaces([])
+    |> add_lines([])
+    |> add_separators
+    |> insert_new_lines([])
   end
 
-  def insert_new_lines_into_board(formatted_rows,[head | tail]) do
-    List.insert_at(formatted_rows, -1,  head ++ ["\n"]) 
-    |> insert_new_lines_into_board(tail)
+  defp add_spaces([], accumulator), do: accumulator
+
+  defp add_spaces([head|tail], accumulator) do
+    spaced_row = add_spaces_to_tiles_trailing(head)
+    leading_spaced_row = add_spaces_to_tiles_leading(spaced_row)
+    list = List.insert_at(accumulator, 100, leading_spaced_row)
```

Comment:
@gemcfadyen it's essentially an `append(...)`. `List.insert_at(...)` wil add stuff at the end of the list if the `at` is too large.